# varscan-hmm

A SNP and indel variant caller using Hidden Markov Models (HMMs) to better account for positional independence and linkage disequilibrium between genetic loci.

## Table of Contents

- [Contributors](#contributors)
- [So...what is a Hidden Markov Model? (Methodology)](#sowhat-is-a-hidden-markov-model-methodology)
- [How do I run it?](#how-do-i-run-it)
  - [Requirements](#requirements)
  - [Installation](#installation)
  - [Usage](#usage)
  - [File Formats](#file-formats)
- [Why should I run it? (Benchmarking Analysis)](#why-should-i-run-it-benchmarking-analysis)

## Contributors

- Gary Lin (A16915179)
- Lorenzo Olmo Marchal (A17013640)
- Nabeeha Rashid (A16851960)

We aim to implement the methods outlined by Tang, et. al (2019) in their [publication](https://humgenomics.biomedcentral.com/articles/10.1186/s40246-019-0194-6) in *Human Genomics*, open access of which is provided under the [Creative Commons Attribution 4.0 International License](http://creativecommons.org/licenses/by/4.0/). This project is intended to be our final project in CSE 185: Advanced Bioinformatics Laboratory for the Spring 2024 quarter, to be presented to our peers and Professor Melissa Gyrmrek.

<cite>
Tang, M., Hasan, M., Zhu, H. et al. vi-HMM: a novel HMM-based method for sequence variant identification in short-read data. Hum Genomics 13, 9 (2019). https://doi.org/10.1186/s40246-019-0194-6
</cite>

## So...what is a Hidden Markov Model? (Methodology)

TODO:

- Explain what a HMM is
  - Explain what the Viterbi Algorithm is
- Explain what HMMs are used for in this paper
- if you can find animations that'd be great

## How do I run it?

### Requirements

### Installation

### Usage

You need a mpileup file

### File Formats

## Why should I run it? (Benchmarking Analysis)
